Dependency exclusions on extensions are ignored

**Describe the bug**
If for some reason we need to exclude a transitive dependency of a Quarkus extension, the dependency is excluded from the runtime dependencies but not from the deployment dependencies.

**Expected behavior**
The dependency should be excluded from the deployment dependencies too. 

**Actual behavior**
The dependency is not excluded from the deployment dependencies. 

**Concrete use case**

In my project, I need the extension `quarkus-resteasy-multipart` but I want to exclude `resteasy-client` from the extension  (because I don't need it on server side, maybe it should be declared as optional?) so I added the next dependency to my pom file:

```
<dependency>
    <groupId>io.quarkus</groupId>
    <artifactId>quarkus-resteasy-multipart</artifactId>
    <exclusions>
        <exclusion>
            <groupId>org.jboss.resteasy</groupId>
            <artifactId>resteasy-client</artifactId>
        </exclusion>
    </exclusions>
</dependency>
```
When I launch my test, I end up with the next error:
```
java.lang.RuntimeException: java.lang.ExceptionInInitializerError
Caused by: java.lang.ExceptionInInitializerError
Caused by: java.lang.RuntimeException: Failed to start quarkus
Caused by: java.lang.RuntimeException: java.lang.ClassNotFoundException: org.jboss.resteasy.client.jaxrs.internal.CompletionStageRxInvokerProvider
Caused by: java.lang.ClassNotFoundException: org.jboss.resteasy.client.jaxrs.internal.CompletionStageRxInvokerProvider
```

This error is due to the fact that `resteasy-client` is still seen as a deployment dependency so the `ResteasyCommonProcessor` [detects its `Provider`](https://github.com/quarkusio/quarkus/blob/master/extensions/resteasy-common/deployment/src/main/java/io/quarkus/resteasy/common/deployment/ResteasyCommonProcessor.java#L175-L176) defined in SPI mode but at Runtime the dependency is excluded as expected so we get a `ClassNotFoundException`.

--- a/core/deployment/src/test/java/io/quarkus/deployment/runnerjar/ExcludeExtensionDepsTest.java
+++ b/core/deployment/src/test/java/io/quarkus/deployment/runnerjar/ExcludeExtensionDepsTest.java
@@ -1,0 +1,88 @@
+package io.quarkus.deployment.runnerjar;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import io.quarkus.bootstrap.model.AppArtifact;
+import io.quarkus.bootstrap.model.AppDependency;
+import io.quarkus.bootstrap.model.AppModel;
+import io.quarkus.bootstrap.resolver.TsArtifact;
+import io.quarkus.bootstrap.resolver.TsDependency;
+import io.quarkus.bootstrap.resolver.TsQuarkusExt;
+
+/**
+ * The test allowing to make sure that if we exclude an extension from another extension, the
+ * exclusion will be also applied to the deployment dependencies.
+ */
+public class ExcludeExtensionDepsTest extends ExecutableOutputOutcomeTestBase {
+
+    @Override
+    protected TsArtifact modelApp() {
+
+        final TsArtifact extADep1 = TsArtifact.jar("ext-a-dep-1");
+        final TsArtifact extADep2 = TsArtifact.jar("ext-a-dep-2");
+        final TsArtifact extBDep1 = TsArtifact.jar("ext-b-dep-1");
+        addToExpectedLib(extBDep1);
+        final TsArtifact extBDep2 = TsArtifact.jar("ext-b-dep-2");
+        addToExpectedLib(extBDep2);
+        final TsArtifact extBDepTrans1 = TsArtifact.jar("ext-b-dep-trans-1");
+        addToExpectedLib(extBDepTrans1);
+        final TsArtifact extBDepTrans2 = TsArtifact.jar("ext-b-dep-trans-2");
+        addToExpectedLib(extBDepTrans2);
+        final TsArtifact depToExclude1 = TsArtifact.jar("ext-dep-exclude-1");
+        final TsArtifact depToExclude2 = TsArtifact.jar("ext-dep-exclude-2");
+        final TsArtifact depToExclude3 = TsArtifact.jar("ext-dep-exclude-3");
+        final TsArtifact depToExclude4 = TsArtifact.jar("ext-dep-exclude-4");
+        final TsArtifact depToExclude5 = TsArtifact.jar("ext-dep-exclude-5");
+        final TsArtifact depToExclude6 = TsArtifact.jar("ext-dep-exclude-6");
+        extBDepTrans2.addDependency(new TsDependency(extBDep2));
+        extBDepTrans2.addDependency(new TsDependency(depToExclude6));
+        extBDepTrans1.addDependency(new TsDependency(extBDepTrans2));
+        extBDepTrans1.addDependency(new TsDependency(depToExclude2));
+        extBDepTrans1.addDependency(new TsDependency(depToExclude5));
+        extBDep1.addDependency(extBDepTrans1, depToExclude5);
+        extBDep1.addDependency(new TsDependency(depToExclude1));
+
+        final TsQuarkusExt extA = new TsQuarkusExt("ext-a");
+        extA.getRuntime()
+                .addDependency(extADep1);
+        extA.getDeployment()
+                .addDependency(extADep2);
+        final TsQuarkusExt extB = new TsQuarkusExt("ext-b");
+        addToExpectedLib(extB.getRuntime());
+        extB.getRuntime()
+                .addDependency(extBDep1, depToExclude6)
+                .addDependency(extA)
+                .addDependency(depToExclude4);
+        extB.getDeployment()
+                .addDependency(depToExclude3);
+
+        return TsArtifact.jar("app")
+                .addManagedDependency(platformDescriptor())
+                .addManagedDependency(platformProperties())
+                .addDependency(extB, extA.getRuntime(), depToExclude1, depToExclude2, depToExclude3, depToExclude4);
+    }
+
+    @Override
+    protected void assertAppModel(AppModel appModel) throws Exception {
+        final Set<AppDependency> expectedDeployDeps = new HashSet<>();
+        expectedDeployDeps
+                .add(new AppDependency(new AppArtifact("io.quarkus.bootstrap.test", "ext-b-deployment", "1"), "compile"));
+        assertEquals(expectedDeployDeps, new HashSet<>(appModel.getDeploymentDependencies()));
+        final Set<AppDependency> expectedRuntimeDeps = new HashSet<>();
+        expectedRuntimeDeps.add(new AppDependency(new AppArtifact("io.quarkus.bootstrap.test", "ext-b", "1"), "compile"));
+        expectedRuntimeDeps.add(new AppDependency(new AppArtifact("io.quarkus.bootstrap.test", "ext-b-dep-1", "1"), "compile"));
+        expectedRuntimeDeps.add(new AppDependency(new AppArtifact("io.quarkus.bootstrap.test", "ext-b-dep-2", "1"), "compile"));
+        expectedRuntimeDeps
+                .add(new AppDependency(new AppArtifact("io.quarkus.bootstrap.test", "ext-b-dep-trans-1", "1"), "compile"));
+        expectedRuntimeDeps
+                .add(new AppDependency(new AppArtifact("io.quarkus.bootstrap.test", "ext-b-dep-trans-2", "1"), "compile"));
+        assertEquals(expectedRuntimeDeps, new HashSet<>(appModel.getUserDependencies()));
+        final Set<AppDependency> expectedFullDeps = new HashSet<>();
+        expectedFullDeps.addAll(expectedDeployDeps);
+        expectedFullDeps.addAll(expectedRuntimeDeps);
+        assertEquals(expectedFullDeps, new HashSet<>(appModel.getFullDeploymentDeps()));
+    }
+}

--- a/core/deployment/src/test/java/io/quarkus/deployment/runnerjar/ExcludeLibDepsTest.java
+++ b/core/deployment/src/test/java/io/quarkus/deployment/runnerjar/ExcludeLibDepsTest.java
@@ -1,0 +1,80 @@
+package io.quarkus.deployment.runnerjar;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import io.quarkus.bootstrap.model.AppArtifact;
+import io.quarkus.bootstrap.model.AppDependency;
+import io.quarkus.bootstrap.model.AppModel;
+import io.quarkus.bootstrap.resolver.TsArtifact;
+import io.quarkus.bootstrap.resolver.TsDependency;
+import io.quarkus.bootstrap.resolver.TsQuarkusExt;
+
+/**
+ * The test allowing to make sure that if we exclude a library from an extension, the
+ * exclusion will be also applied to the deployment dependencies.
+ */
+public class ExcludeLibDepsTest extends ExecutableOutputOutcomeTestBase {
+
+    @Override
+    protected TsArtifact modelApp() {
+
+        final TsArtifact extADep1 = TsArtifact.jar("ext-a-dep-1");
+        addToExpectedLib(extADep1);
+        final TsArtifact extADep2 = TsArtifact.jar("ext-a-dep-2");
+        addToExpectedLib(extADep2);
+        final TsArtifact extADepTrans1 = TsArtifact.jar("ext-a-dep-trans-1");
+        addToExpectedLib(extADepTrans1);
+        final TsArtifact extADepTrans2 = TsArtifact.jar("ext-a-dep-trans-2");
+        addToExpectedLib(extADepTrans2);
+        final TsArtifact depToExclude1 = TsArtifact.jar("ext-dep-exclude-1");
+        final TsArtifact depToExclude2 = TsArtifact.jar("ext-dep-exclude-2");
+        final TsArtifact depToExclude3 = TsArtifact.jar("ext-dep-exclude-3");
+        final TsArtifact depToExclude4 = TsArtifact.jar("ext-dep-exclude-4");
+        final TsArtifact depToExclude5 = TsArtifact.jar("ext-dep-exclude-5");
+        final TsArtifact depToExclude6 = TsArtifact.jar("ext-dep-exclude-6");
+        extADepTrans2.addDependency(new TsDependency(extADep2));
+        extADepTrans2.addDependency(new TsDependency(depToExclude6));
+        extADepTrans1.addDependency(new TsDependency(extADepTrans2));
+        extADepTrans1.addDependency(new TsDependency(depToExclude2));
+        extADepTrans1.addDependency(new TsDependency(depToExclude5));
+        extADep1.addDependency(extADepTrans1, depToExclude5);
+        extADep1.addDependency(new TsDependency(depToExclude1));
+
+        final TsQuarkusExt extA = new TsQuarkusExt("ext-a");
+        addToExpectedLib(extA.getRuntime());
+        extA.getRuntime()
+                .addDependency(extADep1, depToExclude6)
+                .addDependency(depToExclude3);
+        extA.getDeployment()
+                .addDependency(depToExclude4);
+
+        return TsArtifact.jar("app")
+                .addManagedDependency(platformDescriptor())
+                .addManagedDependency(platformProperties())
+                .addDependency(extA, depToExclude1, depToExclude2, depToExclude3, depToExclude4);
+    }
+
+    @Override
+    protected void assertAppModel(AppModel appModel) throws Exception {
+        final Set<AppDependency> expectedDeployDeps = new HashSet<>();
+        expectedDeployDeps
+                .add(new AppDependency(new AppArtifact("io.quarkus.bootstrap.test", "ext-a-deployment", "1"), "compile"));
+        assertEquals(expectedDeployDeps, new HashSet<>(appModel.getDeploymentDependencies()));
+        final Set<AppDependency> expectedRuntimeDeps = new HashSet<>();
+        expectedRuntimeDeps.add(new AppDependency(new AppArtifact("io.quarkus.bootstrap.test", "ext-a", "1"), "compile"));
+        expectedRuntimeDeps.add(new AppDependency(new AppArtifact("io.quarkus.bootstrap.test", "ext-a-dep-1", "1"), "compile"));
+        expectedRuntimeDeps.add(new AppDependency(new AppArtifact("io.quarkus.bootstrap.test", "ext-a-dep-2", "1"), "compile"));
+        expectedRuntimeDeps
+                .add(new AppDependency(new AppArtifact("io.quarkus.bootstrap.test", "ext-a-dep-trans-1", "1"), "compile"));
+        expectedRuntimeDeps
+                .add(new AppDependency(new AppArtifact("io.quarkus.bootstrap.test", "ext-a-dep-trans-2", "1"), "compile"));
+        assertEquals(expectedRuntimeDeps, new HashSet<>(appModel.getUserDependencies()));
+        final Set<AppDependency> expectedFullDeps = new HashSet<>();
+        expectedFullDeps.addAll(expectedDeployDeps);
+        expectedFullDeps.addAll(expectedRuntimeDeps);
+        assertEquals(expectedFullDeps, new HashSet<>(appModel.getFullDeploymentDeps()));
+    }
+}

--- a/independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/resolver/TsArtifact.java
+++ b/independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/resolver/TsArtifact.java
@@ -7,6 +7,7 @@ import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.Properties;
+import java.util.function.Supplier;
 import org.apache.maven.model.DependencyManagement;
 import org.apache.maven.model.Model;
 import org.apache.maven.model.Profile;
@@ -112,16 +113,28 @@ public class TsArtifact {
         return addDependency(new TsDependency(dep));
     }
 
+    public TsArtifact addDependency(TsArtifact dep, TsArtifact... excludes) {
+        return addDependency(new TsDependency(dep).exclude(excludes));
+    }
+
     public TsArtifact addDependency(TsQuarkusExt dep) {
         return addDependency(dep, false);
     }
 
     public TsArtifact addDependency(TsQuarkusExt dep, boolean optional) {
+        return addDependency(dep, () -> new TsDependency(dep.getRuntime(), optional));
+    }
+
+    public TsArtifact addDependency(TsQuarkusExt dep, TsArtifact... excludes) {
+        return addDependency(dep, () -> new TsDependency(dep.getRuntime(), false).exclude(excludes));
+    }
+
+    private TsArtifact addDependency(TsQuarkusExt dep, Supplier<TsDependency> dependencyFactory) {
         if (extDeps.isEmpty()) {
             extDeps = new ArrayList<>(1);
         }
         extDeps.add(dep);
-        return addDependency(new TsDependency(dep.getRuntime(), optional));
+        return addDependency(dependencyFactory.get());
     }
 
     public TsArtifact addDependency(TsDependency dep) {

--- a/independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/resolver/TsDependency.java
+++ b/independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/resolver/TsDependency.java
@@ -51,6 +51,13 @@ public class TsDependency {
         return this;
     }
 
+    public TsDependency exclude(TsArtifact... artifacts) {
+        for (TsArtifact artifact : artifacts) {
+            exclude(artifact);
+        }
+        return this;
+    }
+
     public Dependency toPomDependency() {
         final Dependency dep = new Dependency();
         dep.setGroupId(artifact.groupId);

--- a/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/DeploymentInjectingDependencyVisitor.java
+++ b/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/DeploymentInjectingDependencyVisitor.java
@@ -12,6 +12,7 @@ import java.nio.file.FileSystem;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
 import java.util.Properties;
@@ -20,6 +21,7 @@ import org.eclipse.aether.artifact.Artifact;
 import org.eclipse.aether.artifact.DefaultArtifact;
 import org.eclipse.aether.graph.Dependency;
 import org.eclipse.aether.graph.DependencyNode;
+import org.eclipse.aether.graph.Exclusion;
 import org.eclipse.aether.repository.RemoteRepository;
 import org.eclipse.aether.util.artifact.JavaScopes;
 import org.jboss.logging.Logger;
@@ -80,7 +82,8 @@ public class DeploymentInjectingDependencyVisitor {
         collectRuntimeExtensions(root.getChildren());
         // resolve and inject deployment dependencies
         for (DependencyNode rtNode : runtimeNodes) {
-            replaceWith(rtNode, collectDependencies((Artifact) rtNode.getData().get(QUARKUS_DEPLOYMENT_ARTIFACT)));
+            replaceWith(rtNode, collectDependencies((Artifact) rtNode.getData().get(QUARKUS_DEPLOYMENT_ARTIFACT),
+                    rtNode.getDependency().getExclusions()));
         }
     }
 
@@ -169,11 +172,14 @@ public class DeploymentInjectingDependencyVisitor {
         originalNode.setChildren(children);
     }
 
-    private DependencyNode collectDependencies(Artifact artifact) throws BootstrapDependencyProcessingException {
+    private DependencyNode collectDependencies(Artifact artifact, Collection<Exclusion> exclusions)
+            throws BootstrapDependencyProcessingException {
         try {
-            return managedDeps.isEmpty() ? resolver.collectDependencies(artifact, Collections.emptyList(), mainRepos).getRoot()
+            return managedDeps.isEmpty()
+                    ? resolver.collectDependencies(artifact, Collections.emptyList(), mainRepos, exclusions).getRoot()
                     : resolver
-                            .collectManagedDependencies(artifact, Collections.emptyList(), managedDeps, mainRepos, "test",
+                            .collectManagedDependencies(artifact, Collections.emptyList(), managedDeps, mainRepos, exclusions,
+                                    "test",
                                     "provided")
                             .getRoot();
         } catch (AppModelResolverException e) {

--- a/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenArtifactResolver.java
+++ b/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenArtifactResolver.java
@@ -24,6 +24,7 @@ import org.eclipse.aether.collection.CollectResult;
 import org.eclipse.aether.collection.DependencyCollectionException;
 import org.eclipse.aether.graph.DefaultDependencyNode;
 import org.eclipse.aether.graph.Dependency;
+import org.eclipse.aether.graph.Exclusion;
 import org.eclipse.aether.impl.RemoteRepositoryManager;
 import org.eclipse.aether.installation.InstallRequest;
 import org.eclipse.aether.installation.InstallationException;
@@ -224,7 +225,12 @@ public class MavenArtifactResolver {
 
     public CollectResult collectDependencies(Artifact artifact, List<Dependency> deps, List<RemoteRepository> mainRepos)
             throws BootstrapMavenException {
-        final CollectRequest request = newCollectRequest(artifact, mainRepos);
+        return collectDependencies(artifact, deps, mainRepos, Collections.emptyList());
+    }
+
+    public CollectResult collectDependencies(Artifact artifact, List<Dependency> deps, List<RemoteRepository> mainRepos,
+            Collection<Exclusion> exclusions) throws BootstrapMavenException {
+        final CollectRequest request = newCollectRequest(artifact, mainRepos, exclusions);
         request.setDependencies(deps);
         try {
             return repoSystem.collectDependencies(repoSession, request);
@@ -284,10 +290,11 @@ public class MavenArtifactResolver {
     }
 
     public CollectResult collectManagedDependencies(Artifact artifact, List<Dependency> deps, List<Dependency> managedDeps,
-            List<RemoteRepository> mainRepos, String... excludedScopes) throws BootstrapMavenException {
+            List<RemoteRepository> mainRepos, Collection<Exclusion> exclusions, String... excludedScopes)
+            throws BootstrapMavenException {
         try {
             return repoSystem.collectDependencies(repoSession,
-                    newCollectManagedRequest(artifact, deps, managedDeps, mainRepos, excludedScopes));
+                    newCollectManagedRequest(artifact, deps, managedDeps, mainRepos, exclusions, excludedScopes));
         } catch (DependencyCollectionException e) {
             throw new BootstrapMavenException("Failed to collect dependencies for " + artifact, e);
         }
@@ -295,6 +302,12 @@ public class MavenArtifactResolver {
 
     private CollectRequest newCollectManagedRequest(Artifact artifact, List<Dependency> deps, List<Dependency> managedDeps,
             List<RemoteRepository> mainRepos, String... excludedScopes) throws BootstrapMavenException {
+        return newCollectManagedRequest(artifact, deps, managedDeps, mainRepos, Collections.emptyList(), excludedScopes);
+    }
+
+    private CollectRequest newCollectManagedRequest(Artifact artifact, List<Dependency> deps, List<Dependency> managedDeps,
+            List<RemoteRepository> mainRepos, Collection<Exclusion> exclusions, String... excludedScopes)
+            throws BootstrapMavenException {
         final List<RemoteRepository> aggregatedRepos = aggregateRepositories(mainRepos, remoteRepos);
         final ArtifactDescriptorResult descr = resolveDescriptorInternal(artifact, aggregatedRepos);
         Collection<String> excluded;
@@ -335,11 +348,16 @@ public class MavenArtifactResolver {
             }
         }
 
-        return new CollectRequest()
-                .setRootArtifact(artifact)
+        final CollectRequest request = new CollectRequest()
                 .setDependencies(mergeDeps(deps, originalDeps, managedVersions))
                 .setManagedDependencies(mergedManagedDeps)
                 .setRepositories(aggregateRepositories(aggregatedRepos, newResolutionRepositories(descr.getRepositories())));
+        if (exclusions.isEmpty()) {
+            request.setRootArtifact(artifact);
+        } else {
+            request.setRoot(new Dependency(artifact, JavaScopes.COMPILE, false, exclusions));
+        }
+        return request;
     }
 
     public List<RemoteRepository> newResolutionRepositories(List<RemoteRepository> repos) {
@@ -359,10 +377,14 @@ public class MavenArtifactResolver {
         }
     }
 
-    private CollectRequest newCollectRequest(Artifact artifact, List<RemoteRepository> mainRepos)
-            throws BootstrapMavenException {
+    private CollectRequest newCollectRequest(Artifact artifact, List<RemoteRepository> mainRepos) {
+        return newCollectRequest(artifact, mainRepos, Collections.emptyList());
+    }
+
+    private CollectRequest newCollectRequest(Artifact artifact, List<RemoteRepository> mainRepos,
+            Collection<Exclusion> exclusions) {
         return new CollectRequest()
-                .setRoot(new Dependency(artifact, JavaScopes.RUNTIME))
+                .setRoot(new Dependency(artifact, JavaScopes.RUNTIME, false, exclusions))
                 .setRepositories(aggregateRepositories(mainRepos, remoteRepos));
     }
 

